File: Navigation/Navigation/Navigation/Private/Map.cpp
#include "Map.h"


template<typename G>
Map<G>::Map()
{

}

template<typename G>
Map<G>::~Map()
{

}

template<typename G>
inline int Map<G>::GetIndex(int w, int x, int y)
{
	return x + y * w;
}

template<typename G>
inline IMap<G> * Map<G>::Init(int w, int h, std::unordered_map<int, int> height, std::unordered_map<int, int> flag)
{
	_w = w;
	_h = h;
	_grounds.reserve(_w * _h);
	for (int y = 0; y < _h; y++)
	{
		for (int x = 0; x < _w; x++)
		{
			auto ground = std::shared_ptr<IGround>(new G());
			int index = GetIndex(x, y);
			ground->Init(x, y, height[index], flag[index]);
			_grounds[index] = ground;
		}
	}
	return this;
}

template<typename G>
IMap<G> * Map<G>::AnalysisMap()
{
	return this;
}

template<typename G>
inline std::shared_ptr<G> Map<G>::GetGround(int index)
{
	return _grounds[index];
}

template<typename G>
inline int Map<G>::GetWidth() const
{
	return _w;
}

template<typename G>
inline int Map<G>::GetHeight() const
{
	return _h;
}

template<typename G>
inline int Map<G>::GetIndex(int x, int y) const
{
	return x + y * _w;
}

template<typename G>
inline int Map<G>::GetX(int index) const
{
	return index % _w;
}

template<typename G>
inline int Map<G>::GetY(int index) const
{
	return index / _w;
}

File: SearchPath/Map.lua
local Ground = require('SearchPath.Ground')
local Map = Class('Map')

function Map:Map()
    self.width = 0
    self.height = 0
    self.grounds = {}
    return self
end

function Map:GetIndex(x, y)
    if x > 0 and x <= self.width and y > 0 and y <= self.height then
        return (y - 1) * self.width + x
    else
        return -1
    end
end

function Map:GetPos(index)
    local x = index % self.width
    x = x == 0 and self.width or x
    local y = math.floor((index - 1) / self.width) + 1
    return x, y
end

function Map:Init(width, height)
    self.width = width
    self.height = height
    for i = 1, self.width * self.height do
        local x, y = self:GetPos(i)
        table.insert(self.grounds, Ground.New():Init(x, y, math.floor(math.random() + 0.2)))
    end
    return self
end

function Map:GetGround(index)
    return self.grounds[index]
end

return Map